src/vm_pmu.rs

```diff
@@ -368,4 +373,149 @@ impl VmPmuState {
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::configure_matching_counters(
+                counter_index,
+                counter_mask,
+                config_flags,
+                event_type,
+                event_data,
+            )
+        } else {
+            fn is_firmware_counter(event_type: PmuEventType) -> bool {
+                matches!(event_type, PmuEventType::Firmware(_))
+            }
+
+            fn is_fixed_counter(counter_index: usize) -> bool {
+                // CYCLE and INSRET are fixed counters
+                counter_index == 0 || counter_index == 2
+            }
+
+            fn is_timer(counter_index: u64) -> bool {
+                counter_index == 1
+            }
+
+            if is_firmware_counter(event_type) {
+                return Err(SbiError::InvalidParam);
+            }
+
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let index = if !config_flags.is_skip_match() {
+                let mut bitmask_iter = CounterMaskIter::new(counter_index, counter_mask);
+                bitmask_iter
+                    .find(|i| {
+                        matches!(self.counter_state[*i], NotConfigured) || is_fixed_counter(*i)
+                    })
+                    .map_or_else(|| Err(SbiError::InvalidParam), |i| Ok(i as u64))
+            } else {
+                if is_timer(counter_index) {
+                    return Err(SbiError::InvalidParam);
+                }
+
+                // If skip_match is set, the counter must already be configured
+                let state = self
+                    .counter_state
+                    .get(counter_index as usize)
+                    .ok_or(SbiError::InvalidParam)?;
+                if matches!(state, Started(_) | Configured(_)) {
+                    Ok(counter_index)
+                } else {
+                    Err(SbiError::InvalidParam)?
+                }
+            }?;
+
+            CSR.scounterselect.set(index);
+            if matches!(
+                event_type,
+                PmuEventType::Cache(_) | PmuEventType::Hardware(_)
+            ) {
+                CSR.shpmevent.read_and_set_bits(event_type.raw());
+            } else {
+                CSR.shpmevent.read_and_set_bits(event_data);
+            }
+            CSR.shpmevent
+                .read_and_set_field(riscv_regs::defs::shpmevent::sinh);
+
+            if config_flags.is_clear_value() {
+                CSR.shpmcounter.set(0);
+            }
+
+            if config_flags.is_auto_start() {
+                CSR.scountinhibit
+                    .set(CSR.scountinhibit.read_and_clear_bits(1 << index));
```

Comment:
You just need the `read_and_clear_bits()` here. As written, this will actually undo the bit-setting.